File: 15.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> results;
        sort(nums.begin(), nums.end());
        int n = nums.size();
        for (int i = 0; i < n-2; ++i) {
            if (i > 0 && nums[i] == nums[i-1]) {
                continue;
            }
            int target = nums[i]*(-1);
            int left = i + 1;
            int right = n - 1;
            while (left < right) {
                if (nums[left] + nums[right] == target) {
                    vector<int> tmp;
                    tmp.push_back(target*(-1));
                    tmp.push_back(nums[left]);
                    tmp.push_back(nums[right]);
                    results.push_back(tmp);
                    while (left < right && nums[left] == nums[left+1]) left++;
                    while (left < right && nums[right] == nums[right -1]) right--;
                    left++;
                    right--;
                } else if (nums[left] + nums[right] <  target) {
                    left ++;
                } else {
                    right--;
                }
            }
        }
        return results;
    }
    
    
};

File: python/two_pointers/15.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        results = []
        for i in range(len(nums)-1):

            if i > 0 and nums[i] == nums[i-1]:
                continue
            self.findTwoSum(i+1, len(nums)-1, nums, -nums[i], results)
        return results
    def findTwoSum(self, left, right, nums, target, results):
        while left < right:
            if nums[left] + nums[right] == target:
                results.append([-target, nums[left], nums[right]])
                left+=1
                right-=1
                while left > 0 and left < right and nums[left] == nums[left-1]:
                    left+=1
                while right < len(nums) - 1 and left < right and nums[right] == nums[right + 1]:
                    right-=1
            elif nums[left] + nums[right] < target:
                left+=1
            else:
                right-=1
        
            
